contracts/blender/src/merkle_tree.rs

```diff
@@ -0,0 +1,90 @@
+use ink_prelude::{boxed::Box, vec, vec::Vec};
+use ink_storage::traits::{PackedAllocate, PackedLayout, SpreadAllocate, SpreadLayout};
+use scale::{Decode, Encode};
+#[cfg(feature = "std")]
+use scale_info::TypeInfo;
+
+/// Abstraction for the elements kept in nodes (both in leaves and inner nodes).
+#[cfg(feature = "std")]
+pub trait TreeElement: Clone + PackedLayout + PackedAllocate + TypeInfo + 'static {}
+#[cfg(feature = "std")]
+impl<T: Clone + PackedLayout + PackedAllocate + TypeInfo + 'static> TreeElement for T {}
+
+#[cfg(not(feature = "std"))]
+pub trait TreeElement: Clone + PackedLayout + PackedAllocate {}
+#[cfg(not(feature = "std"))]
+impl<T: Clone + PackedLayout + PackedAllocate> TreeElement for T {}
+
+/// Abstraction for a two-to-one hashing function.
+///
+/// This is used to compute hash in parent node from the children ones.
+pub trait KinderBlender<H>: PackedLayout + Default {
+    /// Compute hash from two hashes.
+    fn blend_kinder(left: &H, right: &H) -> H;
+}
+
+/// Simplified binary tree that represents a Merkle tree over some set of hashes.
+///
+/// It has `LEAVES` leaves (and thus `2 * LEAVES - 1` nodes in general). `KB` is used for computing
+/// values in the inner nodes. `TE` represents *both* leaves' values and inner nodes' values.
+///
+/// `LEAVES` must be power of `2`.
+#[derive(Clone, Eq, PartialEq, Decode, Encode, PackedLayout, SpreadLayout, SpreadAllocate)]
+#[cfg_attr(feature = "std", derive(TypeInfo, ink_storage::traits::StorageLayout))]
+pub struct MerkleTree<TE: TreeElement, KB: KinderBlender<TE>, const LEAVES: usize> {
+    /// Array of node values (root is at [1], children are in [2n] and [2n+1]).
+    nodes: Vec<TE>,
```

Comment:
This is of course just temporary and more mature implementation will use `Mapping`, correct?

Reply:
why? perfectly I would keep here an array `[TE; 2 * LEAVES]`, but this doesn't work for 2 reasons:
 - expressions on const generics are not supported in Rust yet (still unstable grrr)
 - big arrays don't work with storage spreading/packing/layouting and it's much easier to have a vector

mapping doesn't give us anything valuable I think (unless it can optimize read access for index sequences like `n`/`n/2`/`n/4`/...

I will also summon @DamianStraszak here

Reply:
If my understanding is correct, `Vec<_>` will be stored as a single element always so for reading it will read the whole thing into Wasm's VM. What's the limit of that? Are you sure we will never store more than the system allows for and basically brick the contract?

Reply:
in `Spread` fashion every element will have its own cell (and `Blender` is `SpreadAllocate`) - that's my belief

Reply:
I will go for the mapping; but I will just leave it one thought: initially I was trying to use array (since the size is already known), but some of the memory-related traits are not supported for big arrays

Reply:
done